usr/share/rear/rescue/default/500_ssh.sh

```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+                        LogUserOutput "please enter a strong passphrase which will be used when running 'rear recover'."
+                        LogUserOutput "- Entering an empty passphrase will *not* copy unprotected private keys to the rescue medium."
+                        LogUserOutput "- Entering 'UnProtected' (without quotes) will copy unprotected private keys onto the rescue medium."
+                        LogUserOutput ""
+
+                        while true; do
+                            LogUserOutput "SSH client passphrase (or empty, or 'UnProtected'): \c"
+                            read -s passphrase <&6
+                            UserOutput ""
+
+                            case "$passphrase" in
+                                "")
+                                    LogUserOutput "SSH private keys will not be copied"
+                                    break
+                                    ;;
+
+                                UnProtected)
+                                    LogUserOutput ""
+                                    LogUserOutput "WARNING: This will protentially allow anyone in possession of the rescue medium"
+                                    LogUserOutput "unauthenticated access to other systems."
+                                    LogUserOutput ""
+                                    user_input="$( UserInput -I SSH_PRIVATE_KEYS_UNPROTECTED_CONFIRMATION \
+                                        -t 0 -p "Type 'yes' if you want to allow unauthenticated access" )"
+                                    if [ "$user_input" == "yes" ]; then
+                                        LogUserOutput "SSH private keys will be copied in unprotected form"
+                                        break
+                                    fi
+                                    ;;
+
+                                *)
+                                    LogUserOutput "Repeat SSH client passphrase: \c"
+                                    read -s repeated_passphrase <&6
+                                    UserOutput ""
+                                    if [ "$passphrase" != "$repeated_passphrase" ]; then
+                                        LogUserOutput "Passphrases do not match"
+                                        passphrase=""
+                                    else
+                                        LogUserOutput "SSH private keys will be copied and passphrase-protected"
+                                        break
+                                    fi
+                                    ;;
+                            esac
+                        done
+                    fi
+
+                    # Carry out the user's decision
+                    case "$passphrase" in
+                        "")
+                            Log "Ignoring unprotected private key $original_file"
+                            rm "$target_file"
```

Comment:
First copy and then remove? I'd really prefer forward logic, if possible. Like copy only if it is protected.

Reply:
I did so first. But then we'd change the original private key file by just testing for password-protection. Trying to assign an empty passphrase for testing changes not only the private key file's mtime, but also its contents.

I'd rather leave the original private key file untouched and this seemed the simpler solution. An alternative would be to copy twice, first to some staging area and then into the target directory. Doing so would require more cleanup.

Reply:
Yes, `ssh-keygen` must modify the file. 

I think your entire code will become easier to read and understand if you split the `$passphrase` variable into two proper ReaR variables that are also explained and set in `default.conf`:

1. SSH_KEEP_UNPROTECTED_KEYS=false is the default and can be enabled to allow unprotected keys
1. SSH_UNPROTECTED_KEYS_SET_PASSPHRASE=foobar can be set to protect the keys in the rescue media

Such variables will enable you to write simple if statements that read much better and don't overload the passhphrase variable with multiple meanings.